patches/tModLoader/Terraria/ModLoader/ChestLoot.cs

```diff
@@ -0,0 +1,217 @@
+			candidatesByType = new();
+			candidatesByMatch = new();
+		}
+
+		private static bool CheckInvalidness(int item) {
+			return item < ItemID.None || item >= ItemLoader.ItemCount;
+		}
+
+		private static bool CheckInvalidness(Item item) {
+			ArgumentNullException.ThrowIfNull(item, nameof(item));
+			return CheckInvalidness(item.type);
+		}
+
+		public bool Has(int item) {
+			if (item == 0)
+				return false;
+
+			cache = item;
+			if (CheckInvalidness(item))
+				return false;
+			return items.FindIndex(x => x.item.Equals(cache)) != -1;
+		}
+
+		public bool Has(Item item) {
+			return Has(item.type);
+		}
+
+		public bool Has(Predicate<Item> match) {
+			cache = match;
+			return items.FindIndex(x => ((Predicate<Item>)cache)(x.item)) != -1;
+		}
+
+		public bool Put(int item, Func<bool> condition = null) {
+			return Put(ContentSamples.ItemsByType[item], condition);
+		}
+
+		public bool PutAt(int index, int item, Func<bool> condition = null) {
+			return PutAt(index, ContentSamples.ItemsByType[item], condition);
+		}
+
+		public bool PutAt(Predicate<Item> match, int item, Func<bool> condition = null) {
+			return PutAt(match, ContentSamples.ItemsByType[item], condition);
+		}
+
+		public bool Put(Item item, Func<bool> condition = null) {
+			if (Has(item)) {
+				return false;
+			}
+
+			items.Add(new(item, condition));
+			return true;
+		}
+
+		public bool PutAt(int index, Item item, Func<bool> condition = null) {
+			if (Has(item)) {
+				return false;
+			}
+
+			items.Insert(index, new(item, condition));
+			return true;
+		}
+
+		public bool PutAt(Predicate<Item> match, Item item, Func<bool> condition = null) {
+			cache = match;
+			int index = items.FindIndex(x => ((Predicate<Item>)cache)(x.item));
+			if (CheckInvalidness(item) || index != -1) {
+				return false;
+			}
+
+			items.Insert(index, new(item, condition));
+			return true;
+		}
+
+		public bool Put(Predicate<Item> match, int item, Func<bool> condition = null) {
+			return PutAt(match, item, condition);
+		}
+
+		public bool Put(Predicate<Item> match, Item item, Func<bool> condition = null) {
+			return PutAt(match, item, condition);
+		}
+
+		public bool PutBefore(int destination, int item, Func<bool> condition = null) {
+			return PutAt(destination, item, condition);
+		}
+
+		public bool PutBefore(int destination, Item item, Func<bool> condition = null) {
+			return PutAt(destination, item, condition);
+		}
+
+		public bool PutAfter(int destination, int item, Func<bool> condition = null) {
+			return PutAt(destination + 1, item, condition);
+		}
+
+		public bool PutAfter(int destination, Item item, Func<bool> condition = null) {
+			return PutAt(destination + 1, item, condition);
+		}
+
+		public bool Remove(int item) {
+			cache = item;
+			if (CheckInvalidness(item) || items.FindIndex(group => group.item.type.Equals(cache)) == -1) {
+				return false;
+			}
+
+			candidatesByType.Add(item);
+			return true;
+		}
+
+		public bool Remove(Predicate<Item> match) {
```

Comment:
tML should never allow for removal of content, only disabling/hiding. Otherwise other mods may have their insertion locations messed up